File: bindings/takumi-render-uniffi/jvm/src/main/kotlin/io/github/zhjk/takumi/render/uniffi/TemplateEngine.kt
package io.github.zhjk.takumi.render.uniffi

import io.github.zhjk.takumi.render.uniffi.generated.TakumiRenderUniffiNativeLoader
import io.github.zhjk.takumi.render.uniffi.generated.TemplateEngine as GeneratedTemplateEngine

class TemplateEngine : AutoCloseable {
    private val inner: GeneratedTemplateEngine

    init {
        TakumiRenderUniffiNativeLoader.ensureLoaded()
        inner = GeneratedTemplateEngine()
    }

    fun addSearchPath(path: String) {
        invokeVoid { inner.addSearchPath(path) }
    }

    fun addTemplate(name: String, source: String) {
        invokeVoid { inner.addTemplate(name, source) }
    }

    fun clearTemplates() {
        invokeVoid { inner.clearTemplates() }
    }

    fun render(request: TemplateRequest): String =
        invoke { inner.render(request.toGenerated()) }

    override fun close() {
        inner.close()
    }

    private inline fun invokeVoid(action: () -> Unit) {
        try {
            action()
        } catch (throwable: Throwable) {
            throw translateThrowable(throwable)
        }
    }

    private inline fun <T> invoke(action: () -> T): T =
        try {
            action()
        } catch (throwable: Throwable) {
            throw translateThrowable(throwable)
        }

    private fun translateThrowable(throwable: Throwable): RendererException = when (throwable) {
        is RendererException -> throwable
        else -> RendererException.fromGenerated(throwable)
    }
}

File: bindings/takumi-render-uniffi/csharp/Takumi.Render.UniFFI/TemplateEngine.cs
using Generated = Takumi.Render.UniFFI.Generated;

namespace Takumi.Render.UniFFI;

public sealed class TemplateEngine : IDisposable
{
    private readonly Generated.TemplateEngine _inner;

    public TemplateEngine()
    {
        _inner = new Generated.TemplateEngine();
    }

    public void AddSearchPath(string path)
    {
        Invoke(() => _inner.AddSearchPath(path));
    }

    public void AddTemplate(string name, string source)
    {
        Invoke(() => _inner.AddTemplate(name, source));
    }

    public void ClearTemplates()
    {
        Invoke(_inner.ClearTemplates);
    }

    public string Render(RenderTemplateRequest request)
    {
        return Invoke(() => _inner.Render(request.ToGenerated()));
    }

    public void Dispose()
    {
        _inner.Dispose();
    }

    private static void Invoke(Action action)
    {
        try
        {
            action();
        }
        catch (Exception exception)
            when (exception is Generated.RendererException or RendererException)
        {
            throw TranslateException(exception);
        }
    }

    private static T Invoke<T>(Func<T> action)
    {
        try
        {
            return action();
        }
        catch (Exception exception)
            when (exception is Generated.RendererException or RendererException)
        {
            throw TranslateException(exception);
        }
    }

    private static RendererException TranslateException(Exception exception) =>
        exception switch
        {
            RendererException rendererException => rendererException,
            Generated.RendererException generated => RendererException.FromGenerated(generated),
            _ => new RendererException(RendererErrorKind.Unknown, exception.Message, exception),
        };
}
